File: backends/php-laravel/app/Support/Validator.php
<?php

namespace App\Support;

class Validator
{
    /**
     * @var list<array{field: string, messageKey: string}>
     */
    private array $violations = [];

    public function reject(string $field, string $messageKey): void
    {
        $this->violations[] = ['field' => $field, 'messageKey' => $messageKey];
    }

    public function check(bool $condition, string $field, string $messageKey): void
    {
        if (! $condition) {
            $this->reject($field, $messageKey);
        }
    }

    public function throwIfInvalid(): void
    {
        if ($this->violations !== []) {
            throw new ValidationProblem($this->violations);
        }
    }
}

File: backends/spring-java/src/main/java/dev/stackverse/backend/common/Validator.java
package dev.stackverse.backend.common;

import java.util.ArrayList;
import java.util.List;

/** Collects violations and throws once so all field errors are reported together. */
public class Validator {
    private final List<FieldViolation> violations = new ArrayList<>();

    public void reject(String field, String messageKey) {
        violations.add(new FieldViolation(field, messageKey));
    }

    public void check(boolean condition, String field, String messageKey) {
        if (!condition) {
            reject(field, messageKey);
        }
    }

    public void throwIfInvalid() {
        if (!violations.isEmpty()) {
            throw new ValidationProblem(violations);
        }
    }
}
